cirq-core/cirq/ops/state_preparation_gate.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Quantum gates to prepare a given target state."""
+
+from typing import Any, Dict, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import protocols
+from cirq.ops import raw_types
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class StatePreparationGate(raw_types.Gate):
+    """A unitary qubit gate which resets all qubits to the |0> state
+    and then prepares the target state."""
```

Comment:
@AnimeshSinha1309 Can you please update the comments throughout the PR to reflect that it's implemented as a channel (for reference see the comments on other channels like [ResetChannel](https://github.com/quantumlib/Cirq/blob/578c956b3a9f5d13626b58a3993c084bdb52a6d5/cirq-core/cirq/ops/common_channels.py#L681)). You should also include in comments that compilers should provide custom unitary implementations for state preparation. 

I'm also not sure anymore whether we should call it `StatePreparationGate` or rename it to `StatePreparationChannel` / `ReplacementChannel` ?  In my mind, `StatePreparationGate` should encourage the intended use of compilers replacing it with custom unitary implementations. cc @viathor I'd be curious to know your thoughts on this

Reply:
+1 for `StatePreparationChannel`, that's a good point.

Reply:
Yeah, refactored.